package com.appoint.controller;

import java.io.InvalidClassException;
import java.util.List;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.CrossOrigin;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

import com.appoint.exception.CustomException;
import com.appoint.model.Appointment;
import com.appoint.service.AppointmentService;

@RestController
@RequestMapping("/appointment")
@CrossOrigin(origins = "*", maxAge = 3600)
public class AppointmentController {
	
	@Autowired
	private AppointmentService appointmentService;
	
	@PostMapping("/created/{doctorId}/{name}")
	public ResponseEntity<Appointment> createAppointment(@PathVariable String doctorId, @PathVariable String name, @RequestBody Appointment appointment)  throws CustomException, InvalidClassException{
		
		try {
			Appointment appointment2 = appointmentService.createAppointment(doctorId, name, appointment);
			return ResponseEntity.ok(appointment2);
		} catch (CustomException e) {
			return new ResponseEntity<>(HttpStatus.INTERNAL_SERVER_ERROR);
		}
	}
	
	@GetMapping("/list/{appointId}")
	public ResponseEntity<List<Appointment>> getAppointmentList (@PathVariable String appointId) throws CustomException {
		try {
			List<Appointment> getAppointments = appointmentService.getAppointmentById(appointId);
			return ResponseEntity.ok(getAppointments);
		} catch (CustomException e) {
			return new ResponseEntity<>(HttpStatus.INTERNAL_SERVER_ERROR);
		}
	}
	
	@GetMapping("/listPerUser/{username}")
	public ResponseEntity<List<Appointment>> getAppointmentListUsername (@PathVariable String username) throws CustomException {
		try {
			List<Appointment> getAppointments = appointmentService.getAppointByUsername(username);
			return ResponseEntity.ok(getAppointments);
		} catch (CustomException e) {
			return new ResponseEntity<>(HttpStatus.INTERNAL_SERVER_ERROR);
		}
	}

}
